api/wasm/v0/cpp/proxy_wasm_externs.h

```diff
@@ -0,0 +1,121 @@
+/*
+ * Proxy-WASM ABI.
+ */
+// NOLINT(namespace-envoy)
+
+#pragma once
+
+#include <stddef.h>
+#include <stdint.h>
+
+//
+//  ABI calls from the host into the VM.
+//
+//  These will typically be implemented by a language specific SDK which will provide an API on top
+//  of this ABI e.g. the C++ SDK provides a proxy_wasm_api.h implemenation of the API on top of this
+//  ABI.
+//
+
+// Non-stream calls.
+
+/**
+ * Called when the VM starts for each plugin.
+ * @param root_context_id is an identifier with the lifetime of the VM which will be used for
+ * one or more related plugins and their corresponding proxy_on_configure(), proxy_on_done() and
+ * proxy_on_delete() calls. It is also be provided during creation of stream context for those
+ * plugins.
+ * @param configuration_size is the size of any configuration available via proxy_get_configuration.
```

Comment:
Why do I have to specify a size here? What does that do and why?

Reply:
The ABI provides the size so that you can decide to not read it (e.g. if it is of size 0) or perhaps to signal an error (it is too large).  This is true of the other ABI calls which we will see later, e.g. on_body which makes available a buffer.  Consumers can choose to wait until enough data is available e.g. to parse a header or they can use the provided size to retrieve all more some portion of the available data.

Reply:
I commented on this below, but this seems very convoluted for configuration. Why not just pass the config directly vs. this callback mechanism?

Reply:
This is an ABI.  The SDK loads the configuration and converts to a language appropriate API (e.g. string_view) for the callback.  For the ABI we are trying to be as consistent as possible to make it easy to understand ownership and build a good language specific SDK.  As I said above, one of those consistency decisions is to always have the WASM code retrieve data from the host rather than having it sent in.  Different languages handle memory management differently and the standards around common WASM GC are still under development so consistency in this area is particularly important.

Reply:
Resolved below.